fix: `DeleteAccount` and `AddMigratedKeyPair` declaration change

saledjenic opened the pull request:
Both functions `DeleteAccount` and `AddMigratedKeyPair` require password to be provided in order to delete account from the keystore properly (removing account from the cache and deleting corresponding local keystore file).

Password parameter can be also an empty string, since there are cases when an account is not added to the keystore (in case of keycard account), so we have nothing to delete.

@rasom @qfrank this will be a breaking change for mobile, this is continuation to the changes done here: https://github.com/status-im/status-go/pull/3133

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (16)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 1f509c21 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3151/1/) | 2023-01-31 18:19:53 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230131-181819-1f509c-pr3151.zip) |
| :heavy_check_mark: | 1f509c21 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3151/1/) | 2023-01-31 18:22:50 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230131-181814-1f509c-pr3151.aar) |
| :heavy_check_mark: | 1f509c21 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3151/1/) | 2023-01-31 18:25:15 | ~6 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230131-181815-1f509c-pr3151.zip) |
| :heavy_check_mark: | 1f509c21 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/1/) | 2023-01-31 18:28:19 | ~9 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 17a3755b | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3151/2/) | 2023-01-31 19:27:44 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230131-192630-17a375-pr3151.zip) |
| :heavy_check_mark: | 17a3755b | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3151/2/) | 2023-01-31 19:30:00 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230131-192630-17a375-pr3151.aar) |
| :heavy_check_mark: | 17a3755b | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3151/2/) | 2023-01-31 19:30:13 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230131-192630-17a375-pr3151.zip) |
| :heavy_check_mark: | 17a3755b | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/2/) | 2023-01-31 19:36:31 | ~9 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 9553889e | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3151/3/) | 2023-02-01 08:49:32 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230201-084830-955388-pr3151.zip) |
| :heavy_check_mark: | 9553889e | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/3/) | 2023-02-01 08:50:47 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/3//consoleText) |
| :heavy_check_mark: | 9553889e | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3151/3/) | 2023-02-01 08:52:57 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230201-084826-955388-pr3151.zip) |
| :heavy_check_mark: | 9553889e | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3151/3/) | 2023-02-01 08:53:02 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230201-084826-955388-pr3151.aar) |
| | | | | | | |
| :heavy_check_mark: | c57afb60 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3151/4/) | 2023-02-01 08:50:47 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230201-084937-c57afb-pr3151.zip) |
| :heavy_check_mark: | c57afb60 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3151/4/) | 2023-02-01 08:56:27 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230201-085302-c57afb-pr3151.zip) |
| :heavy_check_mark: | c57afb60 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3151/4/) | 2023-02-01 08:56:43 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230201-085306-c57afb-pr3151.aar) |
| :heavy_check_mark: | c57afb60 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/4/) | 2023-02-01 09:00:26 | ~9 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/4//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 5405180e | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3151/5/) | 2023-02-02 12:39:24 | ~58 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230202-123826-540518-pr3151.zip) |
| :heavy_check_mark: | 5405180e | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3151/5/) | 2023-02-02 12:42:50 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230202-123822-540518-pr3151.zip) |
| :heavy_check_mark: | 5405180e | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3151/5/) | 2023-02-02 12:44:01 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230202-123822-540518-pr3151.aar) |
| :heavy_check_mark: | 5405180e | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/5/) | 2023-02-02 12:47:35 | ~9 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | ace75717 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3151/6/) | 2023-02-02 12:41:19 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230202-123947-ace757-pr3151.zip) |
| :heavy_check_mark: | ace75717 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3151/6/) | 2023-02-02 12:46:10 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230202-124253-ace757-pr3151.zip) |
| :heavy_check_mark: | ace75717 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3151/6/) | 2023-02-02 12:47:31 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230202-124404-ace757-pr3151.aar) |
| :heavy_check_mark: | ace75717 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/6/) | 2023-02-02 12:58:12 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3151/6//consoleText) |

saledjenic committed:
fix: `DeleteAccount` and `AddMigratedKeyPair` declaration change

Both functions `DeleteAccount` and `AddMigratedKeyPair` require password to be provided
in order to delete account from the keystore properly (removing account from the cache and
deleting corresponding local keystore file).

Password parameter can be also an empty string, since there are cases when an account is
not added to the keystore (in case of keycard account), so we have nothing to delete.

richard-ramos reviewed (approved):
For me it looks good, but let's wait for @rasom @cammellos @qfrank review in case this PR breaks the mobile app

rasom commented on services/accounts/accounts.go:
```diff
@@ -429,7 +441,7 @@ func (api *API) VerifyPassword(password string) bool {
 }
 
-func (api *API) AddMigratedKeyPair(ctx context.Context, kcUID string, kpName string, keyUID string, accountAddresses []string, keyStoreDir string) error {
+func (api *API) AddMigratedKeyPair(ctx context.Context, kcUID string, kpName string, keyUID string, accountAddresses []string, password string) error {
```
This method is not used on mobile so no problem here.

saledjenic replied:
Great! ;)

rasom commented on services/accounts/accounts.go:
```diff
@@ -72,23 +72,35 @@ func (api *API) GetAccounts(ctx context.Context) ([]*accounts.Account, error) {
-		err = api.manager.DeleteAccount(address, "")
-		var e *account.ErrCannotLocateKeyFile
-		if err != nil && !errors.As(err, &e) {
+func (api *API) DeleteAccount(ctx context.Context, address types.Address, password string) error {
```
Just for the context: why the account can not be deleted without the password?

saledjenic replied:
@rasom cause that's the way how delete account from the keystore function works, check it here: 
https://github.com/status-im/status-go/blob/develop/vendor/github.com/ethereum/go-ethereum/accounts/keystore/keystore.go#L244-L266

that function is called from here:
https://github.com/status-im/status-go/blob/develop/account/accounts.go#L609-L611

Without provided password we're not able to delete neither keystore file for that account nor keystore cache.

rasom replied:
ok, I se, thanks

rasom commented on services/accounts/accounts.go:
```diff
@@ -441,10 +453,12 @@ func (api *API) AddMigratedKeyPair(ctx context.Context, kcUID string, kpName str
-		_ = api.manager.DeleteAccount(addr, "")
+	if len(password) > 0 {
+		for _, addr := range addresses {
+			// This action deletes an account from the keystore, no need to check for error in this context here, cause if this
+			// action fails from whichever reason the account is still successfully migrated since keystore won't be used any more.
```
I know that this comment is not added specifically in this PR, but I still don't fully understand why it is safe to ignore any error here. Considering we want to remove file, and we do not because of some error, it might be a problem from security pov to not handle this error (or at least to not notify user about it) even if everything was migrated as expected. Am I missing something?

saledjenic replied:
@rasom yes, I agree. That originates from the old `DeleteAccount` function which was not removing cache properly and I was thinking in that time about the case if somebody manually removes the keystore from the file system that we could just simply ignore that and complete the action with success. But I agree, will change that.

saledjenic commented:
@rasom thanks for the review, updated.

rasom reviewed (approved):
It is safe to merge now. UI is not implemented yet on mobile side and it doesn't break CI or anything.
